core/runtime/op_driver/mod.rs

```diff
@@ -45,24 +79,283 @@ pub(crate) trait OpDriver: Default {
-    bool,
-    Result<v8::Local<'s, v8::Value>, v8::Local<'s, v8::Value>>,
-  )>;
+  ) -> Poll<(PromiseId, OpId, OpResult<C>)>;
 
+  /// Return the number of futures currently being polled.
   fn len(&self) -> usize;
 }
+
+#[cfg(test)]
+mod tests {
+  use crate::GetErrorClassFn;
+
+  use super::op_results::*;
+  use super::*;
+  use bit_set::BitSet;
+  use rstest::rstest;
+  use std::future::poll_fn;
+
+  struct TestMappingContext {}
+  impl<'s> OpMappingContextLifetime<'s> for TestMappingContext {
+    type Context = ();
+    type Result = String;
+    type MappingError = anyhow::Error;
+
+    fn map_error(
+      _context: &mut Self::Context,
+      err: Error,
+      _get_error_class_fn: GetErrorClassFn,
+    ) -> UnmappedResult<'s, Self> {
+      Ok(format!("{err:?}"))
+    }
+
+    fn map_mapping_error(
+      _context: &mut Self::Context,
+      err: Self::MappingError,
+    ) -> Self::Result {
+      format!("{err:?}")
+    }
+  }
+  impl OpMappingContext for TestMappingContext {
+    type MappingFn<R: 'static> = for<'s> fn(R) -> Result<String, anyhow::Error>;
+    fn erase_mapping_fn<R: 'static>(f: Self::MappingFn<R>) -> *const fn() {
+      f as _
+    }
+
+    fn unerase_mapping_fn<'s, R: 'static>(
+      f: *const fn(),
+      _context: &mut <Self as OpMappingContextLifetime<'s>>::Context,
+      r: R,
+    ) -> UnmappedResult<'s, Self> {
+      let f: Self::MappingFn<R> = unsafe { std::mem::transmute(f) };
+      f(r)
+    }
+  }
+
+  fn submit_task(
+    driver: &impl OpDriver<TestMappingContext>,
+    scheduling: OpScheduling,
+    id: usize,
+    op: impl Future<Output = i32> + 'static,
+  ) {
+    assert_eq!(
+      None,
+      driver.submit_op_infallible_scheduling(
+        scheduling,
+        1234,
+        id as _,
+        op,
+        |r| { Ok(format!("{r}")) }
+      )
+    );
+  }
+
+  fn submit_task_eager_ready(
+    driver: &impl OpDriver<TestMappingContext>,
+    id: usize,
+    op: impl Future<Output = i32> + 'static,
+    result: i32,
+  ) {
+    assert_eq!(
+      Some(result),
+      driver.submit_op_infallible_scheduling(
+        OpScheduling::Eager,
+        1234,
+        id as _,
+        op,
+        |r| { Ok(format!("{r}")) }
+      )
+    );
+  }
+
+  async fn reap_task(
+    driver: &impl OpDriver<TestMappingContext>,
+    bitset: &mut BitSet,
+    expected: &str,
+  ) {
+    let (promise_id, op_id, result) = poll_fn(|cx| driver.poll_ready(cx)).await;
+    assert!(bitset.insert(promise_id as usize));
+    assert_eq!(1234, op_id);
+    assert_eq!(expected, &(result.unwrap(&mut (), &|_| "Error").unwrap()));
+  }
+
+  #[rstest]
+  #[case::joinset(JoinSetDriver::<TestMappingContext>::default())]
+  #[case::futures_unordered(FuturesUnorderedDriver::<TestMappingContext>::default())]
+  fn test_driver<D: OpDriver<TestMappingContext>>(
+    #[case] driver: D,
+    #[values(2, 16)] count: usize,
+    #[values(OpScheduling::Eager, OpScheduling::Lazy, OpScheduling::Deferred)]
+    scheduling: OpScheduling,
```

Comment:
Side note: I really don't like all this macro magic from `rstest` 😅 

That said, these tests are great.

Reply:
Under the hood it just runs a combination of all those inputs which actually reduces the amount of macro code we need to effectively test all of these cases 😓 

Even though these macros are a bit noisy, the real value is that we can add a lot of code coverage with only a few lines of testing code.